library/public/engine.h

```diff
@@ -0,0 +1,142 @@
+ *
+ * The main class of the libf3d to create all other instances
+ * Configured on creation using an enum, then all objects
+ * can be accessed through their getter.
+ *
+ * Example usage:
+ *
+ * \code{.cpp}
+ *  f3d::engine eng(f3d::engine::CREATE_WINDOW | f3d::engine::CREATE_INTERACTOR);
+ *  f3d::loader& load = eng.getLoader();
+ *  load.addFile("path/to/file");
+ *  load.loadFile(f3d::loader::LoadFileEnum::LOAD_CURRENT);
+ *  f3d::interactor& inter = eng.getInteractor();
+ *  inter.start();
+ * \endcode
+ */
+
+#ifndef f3d_engine_h
+#define f3d_engine_h
+
+#include "export.h"
+#include "window.h"
+
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace f3d
+{
+class options;
+class loader;
+class interactor;
+class F3D_EXPORT engine
+{
+public:
+  class exception : public std::runtime_error
+  {
+  public:
+    exception(const std::string& what = "")
+      : std::runtime_error(what)
+    {
+    }
+  };
+
+  /**
+   * Engine constructor, choose the window type using the enum.
+   * see window.h for details about the window.
+   * When using window::Type::NONE, window and interactor will not be provided by the engine.
+   * When using window::Type::EXTERNAL, interactor will not be provided by the engine.
+   * All objects instances will be created on construction.
+   * Default is window::Type::NATIVE.
+   */
+  explicit engine(window::Type windowType = window::Type::NATIVE);
+
+  /**
+   * Engine destructor, delete all object instances as well
+   */
+  ~engine();
+
+  /**
+   * Engine provide a default options that you can use using engine::getOptions()
+   * But you can use this setter to use other options directly.
+   * It will copy options into engine
+   */
+  engine& setOptions(const options& opt);
+
+  /**
+   * Engine provide a default options that you can use using engine::getOptions()
+   * But you can use this setter to use other options directly.
+   * It will move options into engine
+   */
+  engine& setOptions(options&& opt);
+
+  /**
+   * Get the default options provided by the engine
+   */
+  options& getOptions();
+
+  /**
+   * Get the window provided by the engine, if any.
+   * If not, will throw a engine::exception
+   */
+  window& getWindow();
+
+  /**
+   * Get the loaded provided by the engine
+   */
+  loader& getLoader();
+
+  /**
+   * Get the interactor provided by the engine, if any
+   * If not, will throw a engine::exception
+   */
+  interactor& getInteractor();
+
+  /**
+   * Get a map containing info about the libf3d
+   * Available keys here:
+   * - "Version"
+   * - "Build date"
+   * - "Build system"
+   * - "Compiler"
+   * - "Raytracing module"
+   * - "External rendering module"
+   * - "Exodus module"
+   * - "OpenCASCADE module"
+   * - "Assimp module"
+   * - "Alembic module"
+   * - "VTK version"
+   * - "Copyright_previous"
+   * - "Copyright"
+   * - "License"
+   * - "Authors"
+   */
+  static std::map<std::string, std::string> getLibInfo();
```

Comment:
Any particular reason a string to string map was chosen over a struct? An advantage of a struct is the compiler will tell you when you accidentally type "VKTVersion", and compile time errors are _much_ easier to debug.

Reply:
The idea was to be as flexible as possible. Wdyt @Meakk ?

Reply:
Yes the idea was to be flexible like the options class. However, for this particular map, we could create a new class or an internal struct, it's unlikely to change regularly and most users don't care about it anyway.

Reply:
Thinking back on it, this was just laziness on my part. Adressed here: https://github.com/f3d-app/f3d/pull/360